vispy/color/colormap.py

```diff
@@ -115,45 +118,79 @@ def step(colors, x, controls=None):
     return colors[x_step, ...]
 
 
+# find the first 'controls' value that is smaller than 't'
+def find_color_index(controls=None, t=None, ncolors=None):
+    bn = np.nonzero(controls >= t)
+    j = bn[0][0]-1
+    j = np.clip(j, 0, ncolors-1)
+    return j
+
+
 # GLSL interpolation functions.
-def _glsl_mix(controls=None):
+def _glsl_mix(controls=None, colors=None, texture_map_data=None):
     """Generate a GLSL template function from a given interpolation patterns
-    and control points."""
+    and control points.
+
+    Parameters
+    ----------
+    colors : array-like, shape (n_colors, 4)
+        The control colors used by the colormap.
+        Elements of colors must be convertible to an instance of Color-class.
+
+    controls : list
+        The list of control points for the given colors. It should be
+        an increasing list of floating-point number between 0.0 and 1.0.
+        The first control point must be 0.0. The last control point must be
+        1.0. The number of control points depends on the interpolation scheme.
+
+    texture_map_data : ndarray, shape(LUT_len, 4)
+        Numpy array of size of 1D texture lookup data
+        for luminance to RGBA conversion.
+
+    """
     assert (controls[0], controls[-1]) == (0., 1.)
     ncolors = len(controls)
     assert ncolors >= 2
-    if ncolors == 2:
-        s = "    return mix($color_0, $color_1, t);\n"
-    else:
-        s = ""
-        for i in range(ncolors-1):
-            if i == 0:
-                ifs = 'if (t < %.6f)' % (controls[i+1])
-            elif i == (ncolors-2):
-                ifs = 'else'
-            else:
-                ifs = 'else if (t < %.6f)' % (controls[i+1])
-            adj_t = '(t - %s) / %s' % (controls[i],
-                                       controls[i+1] - controls[i])
-            s += ("%s {\n    return mix($color_%d, $color_%d, %s);\n} " %
-                  (ifs, i, i+1, adj_t))
-    return "vec4 colormap(float t) {\n%s\n}" % s
-
-
-def _glsl_step(controls=None):
+    assert (texture_map_data is not None)
+
+    LUT = texture_map_data
+    LUT_len = texture_map_data.shape[0]
+    LUT_tex_idx = np.linspace(0.0, 1.0, LUT_len)
+
+    for i in range(LUT_len):
+        t = LUT_tex_idx[i]
+        j = find_color_index(controls, t, ncolors)
+        adj_t = (t - controls[j]) / (controls[j+1] - controls[j])
+        LUT[i, 0, :] = _mix_simple(Color(colors[j]).rgba,
+                                   Color(colors[j+1]).rgba, adj_t)
```

Comment:
I may be misunderstanding what this chunk of code is doing, but couldn't this be replaced by `np.interp` in some way (https://docs.scipy.org/doc/numpy/reference/generated/numpy.interp.html)?

At the least, we want to avoid doing all of the `adj_t` and `_mix_simple` stuff individually for each LUT value. With `find_color_index` and the rest of this code the way it is, it's a lot of complexity in python. If we can move it to C via numpy the performance should be much better as things grow.

Lastly, @larsoner @kmuehlbauer your thoughts are appreciated on this too, all of the uses of `linspace` in the colormap code for texture indexes, should they include the end point of 1.0? Right now they do not. Is 1.0 a valid texture coordinate?

Reply:
IIRC yes they span 0 - 1 inclusive

And I agree that `np.interp` should be used if possible

Reply:
@larsoner Beat me about some seconds :+1:

Reply:
Oops, looks like `endpoint=True` is the default. No changes needed regarding linspace. https://docs.scipy.org/doc/numpy/reference/generated/numpy.linspace.html

Reply:
I have to look this up again and again too, I always mix this up with np.arange which has a half open interval.

Reply:
Before commit 49b1a6d850d00932522039ace03c61c3f64b2d17, the code had tests for multiple different types of incoming colors
```
if(isinstance(colors, ColorArray)):
                LUT[i,0,:] = _mix_simple(colors[j].rgba, colors[j+1].rgba, adj_t)
            elif(isinstance(colors[j], str)):
                LUT[i,0,:3] = _mix_simple(ColorArray(colors[j]).rgb, ColorArray(colors[j+1]).rgb, adj_t)
            else:
                LUT[i,0,:len(colors[j])] = _mix_simple(np.array(colors[j]), np.array(colors[j+1]), adj_t)
```
Even that code ran into some issues in some cases. So, changed the code into
`LUT[i,0,:] = _mix_simple(Color(colors[j]).rgba, Color(colors[j+1]).rgba, adj_t)` 
which should work even if Color class is changed to accept new ways to specify the color.